dataflow/src/ops/distinct.rs

```diff
@@ -0,0 +1,252 @@
+use std::collections::HashMap;
+
+use prelude::*;
+
+/// This will get Distinct elements from each group
+/// TODO
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Distinct {
+    // Parent Node
+    src: IndexPair,
+
+    us: Option<IndexPair>,
+
+    group_by: Vec<usize>,
+
+    // A list of records that have been seen
+    seen: HashMap<Vec<DataType>, bool>,
+}
+
+impl Distinct {
+    pub fn new(
+        src: NodeIndex,
+        group_by: Vec<usize>,
+        ) -> Self {
+
+        // TODO add implementation for group by
+        let mut group_by = group_by;
+        group_by.sort();
+        Distinct {
+            src: src.into(),
+            us: None,
+            group_by: group_by,
+            seen: HashMap::new(),
+        }
+    }
+}
+
+impl Ingredient for Distinct {
+    /// Returns a clone of the node
+    fn take(&mut self) -> NodeOperator {
+        Clone::clone(self).into()
+    }
+
+    fn ancestors(&self) -> Vec<NodeIndex> {
+        // I think this is right?
+        vec![self.src.as_global()]
+    }
+
+    fn on_input(
+        &mut self,
+        from: LocalNodeIndex,
+        rs: Records,
+        _: &mut Tracer,
+        _: Option<&[usize]>,
+        _: &DomainNodes,
+        state: &StateMap,
+    ) -> ProcessingResult {
+        debug_assert_eq!(from, *self.src);
+
+        // Check that the records aren't empty
+        if rs.is_empty() {
+            return ProcessingResult {
+                results: rs,
+                misses: vec![],
+            };
+        }
+
+        let us = self.us.unwrap();
+        let db = state
+            .get(&*us)
+            .expect("Distinct must have its own state initialized");
+
+        let mut seen = HashMap::new();
+        let mut cleaned = Vec::new();
+
+        // Preprocess current batch to remove duplicates
+        // since those won't be caught by checking state below.
+        for rec in rs.iter() {
+            let group = rec.iter()
+                .enumerate()
+                .filter_map(|(i, v)| {
+                    if self.group_by.iter().any(|col| col == &i) {
+                        Some(v)
+                    } else {
+                        None
+                    }
+                })
+                .cloned()
+                .collect::<Vec<_>>();
+
+            let positive = rec.is_positive();
+            if seen.contains_key(&group){
+                if !positive{
+                    seen.remove(&group);
+                    cleaned.push(rec.clone());
+                }
+            } else {
+                cleaned.push(rec.clone());
+                seen.insert(group.clone(), true);
+            }
+        }
```

Comment:
I think you could actually make this code a fair amount nicer by just sorting the batch first and then walking it in order. That way you'd know to not skip records whenever the key changes. You can take a look at how we do this [for grouped operators](https://github.com/mit-pdos/distributary/blob/67514028f32d94e78ce1ed2513fb963d94083281/dataflow/src/ops/grouped/mod.rs#L162L174). The implementation for [TopK](https://github.com/mit-pdos/distributary/blob/67514028f32d94e78ce1ed2513fb963d94083281/dataflow/src/ops/topk.rs#L131) is even closer to what you want to do; it also detects where groups start and stop, but the code is a little hairier as a result. Changing the code that way would also eliminate both `seen` and `cleaned` I think.